firehose-source/src/main/java/source/FirehoseSource.java

```diff
@@ -0,0 +1,60 @@
+/*
+ *  Copyright 2015 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package source;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.cloud.stream.annotation.EnableModule;
+import org.springframework.cloud.stream.annotation.Source;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationListener;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.event.ContextRefreshedEvent;
+import org.springframework.integration.websocket.ClientWebSocketContainer;
+import org.springframework.integration.websocket.inbound.WebSocketInboundChannelAdapter;
+import org.springframework.messaging.MessageChannel;
+
+import java.nio.ByteBuffer;
+import java.util.Collections;
+
+/**
+ * @author Vinicius Carvalho
+ */
+@EnableModule(Source.class)
+@Configuration
+public class FirehoseSource implements ApplicationListener<ContextRefreshedEvent>{
+
+    @Autowired
+    private FirehoseProperties metadata;
+
+    @Bean
+    public WebSocketInboundChannelAdapter webSocketInboundChannelAdapter(ClientWebSocketContainer webSocketContainer, ByteBufferMessageConverter converter, MessageChannel output) {
+        WebSocketInboundChannelAdapter adapter = new WebSocketInboundChannelAdapter(webSocketContainer);
+        adapter.setMessageConverters(Collections.singletonList(converter));
+        adapter.setOutputChannel(output);
+        adapter.setPayloadType(ByteBuffer.class);
+        return adapter;
+    }
+
+    @Override
+    public void onApplicationEvent(ContextRefreshedEvent contextRefreshedEvent) {
+        ApplicationContext ctx = contextRefreshedEvent.getApplicationContext();
+        ctx.getBean(ClientWebSocketContainer.class).start();
+        ctx.getBean(WebSocketInboundChannelAdapter.class).start();
```

Comment:
Seems like you found the solution yourself. `@Configuration` classes make
it so that a method call returns the same instance. Could also have
injected the bean in the configuration class itself, amounts to the same,
more or less.
See
http://docs.spring.io/spring/docs/4.2.1.BUILD-SNAPSHOT/spring-framework-reference/htmlsingle/#beans-java-further-information-java-config

Rest assured, I myself learnt things about `@Configuration` not later than
yesterday :)

On Tue, Aug 18, 2015 at 10:50 PM, Vinicius Carvalho <
notifications@github.com> wrote:

> In firehose-source/src/main/java/source/FirehoseSource.java
> <https://github.com/spring-cloud/spring-cloud-stream-modules/pull/3#discussion_r37351604>
> :
>
> > +    private FirehoseProperties metadata;
> > +
> > +    @Bean
> > +    public WebSocketInboundChannelAdapter webSocketInboundChannelAdapter(ClientWebSocketContainer webSocketContainer, ByteBufferMessageConverter converter, MessageChannel output) {
> > +        WebSocketInboundChannelAdapter adapter = new WebSocketInboundChannelAdapter(webSocketContainer);
> > +        adapter.setMessageConverters(Collections.singletonList(converter));
> > +        adapter.setOutputChannel(output);
> > +        adapter.setPayloadType(ByteBuffer.class);
> > +        return adapter;
> > +    }
> > +
> > +    @Override
> > +    public void onApplicationEvent(ContextRefreshedEvent contextRefreshedEvent) {
> > +        ApplicationContext ctx = contextRefreshedEvent.getApplicationContext();
> > +        ctx.getBean(ClientWebSocketContainer.class).start();
> > +        ctx.getBean(WebSocketInboundChannelAdapter.class).start();
>
> @ericbottard <https://github.com/ericbottard> I need some education here.
> PostConstruct can not have any arguments, and I really thought that calling
> wsClient() on a post construct method would return a different instance
> than the one that was set for the container. If not the case I'm changing
> the code, and it was a good lesson learned today.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spring-cloud/spring-cloud-stream-modules/pull/3/files#r37351604>
> .
>